src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -713,15 +707,10 @@ protected internal void OnDocumentsAdded(ImmutableArray<DocumentInfo> documentIn
         /// </summary>
         protected internal void OnDocumentReloaded(DocumentInfo newDocumentInfo)
         {
-            using (_serializationLock.DisposableWait())
```

Comment:
So does this _for now_ still need to acquire this, since any _other_ operation that is still using the serialization lock could still race against this by having it update in the middle?

Reply:
SetCurrentSolution(transform) takes the serialization lock.  it's equivalent to the code we have here.

Reply:
it does:

```
                var newSolution = transformedSolution.WithNewWorkspace(this, currentSolution.WorkspaceVersion + 1);

                Solution oldSolution;
                using (_serializationLock.DisposableWait())
                {
                    oldSolution = Interlocked.CompareExchange(ref _latestSolution, newSolution, currentSolution);
                    if (oldSolution == currentSolution)
                    {
                        // Queue the event but don't execute its handlers on this thread.
                        // Doing so under the serialization lock guarantees the same ordering of the events
                        // as the order of the changes made to the solution.
                        RaiseWorkspaceChangedEventAsync(kind, oldSolution, newSolution, projectId, documentId);

                        return true;
                    }
                }
```

Basically, it produces the new snapshot out of hte lock, but ensures that the update/event is atomic.

Reply:
Ah, I missed that, sorry!

Reply:
No worries :)